source/material_model/reactive_fluid_transport.cc

```diff
@@ -179,28 +202,82 @@ namespace aspect
-                    else if (c == porosity_idx && this->get_timestep_number() > 0)
-                      reaction_rate_out->reaction_rates[q][c] = porosity_change / fluid_reaction_time_scale;
-                    else
-                      reaction_rate_out->reaction_rates[q][c] = 0.0;
-                  }
+                {
+                  const unsigned int bound_fluid_idx = this->introspection().compositional_index_for_name("bound_fluid");
+
+                  const double solid_density = out.densities[q];
+                  const double fluid_density = fluid_out->fluid_densities[q];
+
+                  const double volume_frac_porosity = in.composition[q][porosity_idx];
+                  const double bulk_density = solid_density * (1.0 - volume_frac_porosity) +
+                                              fluid_density * volume_frac_porosity;
+
+                  // We need to convert the porosity, which is a volume fraction, to a mass fraction
+                  // to determine the total mass fraction of water (across both phases). This is then
+                  // used to ensure conservation of water mass during reactions.
+                  const double mass_frac_porosity = volume_frac_porosity * fluid_density / bulk_density;
+
+                  // The total mass fraction of water, which must be conserved after the reactions are done.
+                  const double total_mass_fraction_water = in.composition[q][bound_fluid_idx] * (1 - volume_frac_porosity) * solid_density / bulk_density +
+                                                           mass_frac_porosity;
+
+                  // Determine the change in the mass fraction of porosity from the equilibrium fluid fraction
+                  // and update the mass fraction of porosity
+                  const double mass_frac_porosity_change = (eq_free_fluid_fractions[q] - mass_frac_porosity);
+                  const double new_mass_frac_porosity = mass_frac_porosity + mass_frac_porosity_change;
+
+                  // Since porosity is a volume fraction, convert to volume fraction for determining the reaction rate
+                  const double new_volume_frac_porosity = solid_density * new_mass_frac_porosity /
+                                                          (fluid_density + solid_density * new_mass_frac_porosity - fluid_density * new_mass_frac_porosity);
+
+                  // Determine the new bulk density using the new volume fraction of porosity
+                  const double new_bulk_density = solid_density * (1.0 - new_volume_frac_porosity) +
+                                                  fluid_density * new_volume_frac_porosity;
+
+                  // Determine the bound fluid mass fraction using the new volume fraction of porosity and the original
+                  // mass fraction of water.
+                  const double new_bound_fluid_mass_frac = (total_mass_fraction_water - new_volume_frac_porosity * fluid_density / new_bulk_density) /
```

Comment:
Isn't `new_volume_frac_porosity * fluid_density / new_bulk_density` simply `new_mass_frac_porosity`?
If that's true (and does not change the results) I would change it to `new_mass_frac_porosity` because it's both shorter and easier to read.